### ISSUE TITLE
New run, April 2023

### ISSUE THREAD

**Issue by erikbern** [erikbern/ann-benchmarks#363]
I just ran sift-128-euclidean and here are the results:

(edit: this chart is superseded by a later chart below)

I also ran glove-100-angular, but had a minor bug – see #362 – and I'm rerunning everything just to be on the safe side. Will post it later today.

Ran it on a [r6i.16xlarge](https://instances.vantage.sh/aws/ec2/r6i.16xlarge) with 31 runs in parallel – was able to run everything in just about 8 hours.

Note that these aren't final – I want to give library maintainers a chance to fix any issues. 

This plot is extremely crowded btw – any thoughts on algos we can remove? Or alternatively we could break it up into maybe "libraries" vs "servers" or something

**Comment by erikbern** [comment]
glove-100-angular, 80% complete:

![image](https://user-images.githubusercontent.com/1027979/230752477-4cb2e394-56b7-486e-9124-f5aabd769879.png)

I'm killing this for now since it's $100/day+ to run. But will rerun soon!

**Comment by erikbern** [comment]
Had an 1TB EBS volume but in retrospect that was kind of ridiculous – only needed 34GB up to this point!

**Comment by WPJiang** [comment]
> glove-100-angular, 80% complete:
> 
> ![image](https://user-images.githubusercontent.com/1027979/230752477-4cb2e394-56b7-486e-9124-f5aabd769879.png)
> 
> I'm killing this for now since it's $100/day+ to run. But will rerun soon!
Hi  @erikbern ，in this test result, the recall of glove-100-angular  is close to zero, as is mentioned in  #367 . We checked the code and found that the calculation method of angular distance has changed. The normalized Euclidean distance currently used is order-preserving compared to angular distance,but their values are not equal. In current code，the distance threshold used to determine top-K is directly read from groundtruth (which is angular distance), while the distances of the top-K vectors are  normalized Euclidean distance.  We think this is cause of the nearly zero-recall problem. To solve the problem, we can take out the top-K ID of groundtruth and use the original vector to calculate the threshold for determination. Or use the original annbenchmarks angular distance calculation method.

**Comment by erikbern** [comment]
Ohhh nice catch. I didn't realize changing the distance would matter.

I think the angular distance could be computed from dot products, is that right? Since

`dot(u, v) = |u| * |v| * cos(theta)`

then what we want is I think `theta = arccos( dot(u, v) / (|u| |v|) )`

is that right?

**Comment by benbenwang1999** [comment]
good work！btw, why the result of milvus is better than the original result？

**Comment by erikbern** [comment]
I think Milvus made a lot of improvements since the last run which was multiple years ago

**Comment by benbenwang1999** [comment]
> I think Milvus made a lot of improvements since the last run which was multiple years ago

Thanks a lot, the answer is very helpful. I have another question, why the indicator chooses recall-QPS instead of precision-QPS

**Comment by erikbern** [comment]
> @erikbern ,The angular distance in the current datasets of annbenchmarks is actually the cosine distance, which is calculated as 1 minus the cosine similarity of the vectors. d= 1-cos(u,v)=1-dot(u, v) / (|u| * |v|)，we have committed a new PR #372 to fix it.

Thanks – I also fixed the test in #373 and added an extra test to make sure it's consistent with the dataset

> 
> Besides, would you please consider relaxing the timeout setting to 24 hours? We found that for some datasets, some algorithms (such as NGTqg and qsgNGT) cannot complete the construction within 2 hours, but when the timeout is set to 24 hours, they get good qps-recall performance. Of course, these algorithms’ disadvantage in construction time will be reflected in the Recall-build time performance. 24 hours of construction time is indeed a bit long, but for some offline construction applications, it is acceptable to trade construction time for qps-recall performance.

Responded in #374 but the tl;dr is I don't think we should do that

**Comment by erikbern** [comment]
> I have another question, why the indicator chooses recall-QPS instead of precision-QPS

I think precision and recall is the same thing given the current definition? We ask each algorithm to get the nearest 10 neighbors and we check how many of those are in the real 10 neighbors. Given that those two sets have the same size, I think recall and precision will have the same value.

**Comment by erikbern** [comment]
We run every algorithm 5 times and pick the best run.

**Comment by erikbern** [comment]
Ran glove-100-angular overnight and here are the results:

![image](https://user-images.githubusercontent.com/1027979/232223185-99839672-f43d-444e-872b-26d78cc2eefa.png)

Something seems broken with Redisearch – will look into it (it got zero recall on every query). Weird since it seems to work on smaller tests. Similarly Weaviate's recall isn't great – will run it by them to see if I can improve the configuration.

**Comment by thomasahle** [comment]
Looks like Pynndescent has a bug...

It also looks like it might be interesting to add some more argument options in the "high recall" regime for most algorithms. Only 5 of then currently going above 0.99.

**Comment by erikbern** [comment]
Fixed Redisearch and changed parameters for Weaviate – current results for glove-100-angular

![image](https://user-images.githubusercontent.com/1027979/232317358-1d95b0d5-f0be-41db-8a33-bf036c2070de.png)

@thomasahle agree it would be nice with more high-recall points. Maybe that's the only place Annoy currently seems to have utility :)

**Comment by erikbern** [comment]
Latest results for sift-128-euclidean:

![image](https://user-images.githubusercontent.com/1027979/232342647-d0b7ee27-8fb5-4ec0-9247-144d932fc785.png)

**Comment by lmcinnes** [comment]
Clearly something is going astray with pynndescent for Glove-100. I'll try to take a look and see if I reproduce the issue locally.

**Comment by erikbern** [comment]
> Did fast_pq not work for sift-128-euclidean?

Not sure what happened. It's possible I messed it up somehow. I'm going to re-run it anyway, so hopefully it will work next time (if not then let's fix it)

**Comment by erikbern** [comment]
> Clearly something is going astray with pynndescent for Glove-100. I'll try to take a look and see if I reproduce the issue locally.

This line seems suspicious to me :) https://github.com/erikbern/ann-benchmarks/blob/main/ann_benchmarks/algorithms/pynndescent.py#L33

**Comment by lmcinnes** [comment]
I believe it was intended to be deliberate but it seems like some handling of that in pynndescent has gone astray. That does look like the source of the issue though. I'll try to get something fixed imminently. In the meantime swapping back to cosine should work and just be a little slower.

**Comment by erikbern** [comment]
Ok, nice, sounds like the next run should work then!

**Comment by thomasahle** [comment]
Just realized r6i supports AVX-512, which has float16 support.
Seems [Disk ANN takes advantage of that](https://github.com/erikbern/ann-benchmarks/issues/240) and [ScaNN does too](https://github.com/google-research/google-research/blob/master/scann/scann/hashes/internal/lut16_avx512.inc).
Seems very useful for this kind of workload. I guess all libraries need to support it to be competitive.

**Comment by lmcinnes** [comment]
Out of curiosity are the Jaccard benchmarks going to make it into this run? Since NMSlib Jaccard was added there should be a reasonable number of implementations for the Jaccard benchmark at last.

**Comment by erikbern** [comment]
Yeah I'll re-run all the benchmarks. Haven't verified the Jaccard ones but hopefully they still work :)

**Comment by erikbern** [comment]
Closing this in favor of #399
